### ISSUE TITLE
Internal error at src/full/Agda/TypeChecking/Serialise/Instances/Internal.hs:147

### ISSUE THREAD

**Issue by d-plaindoux** [agda/agda#4869]
Compiler version __2.6.1__ fails with the following code fragment:

```agda
{-# OPTIONS --no-positivity-check #-}

module Sets.Propositions where

open import Data.Nat using (ℕ; zero; suc)

-- true and absurd propositions definition

data ⊤ : Set where
  tt : ⊤

data ⊥ : Set where

data Bad : Set where
  bad : (Bad → ⊥) → Bad

absurd : ⊥
absurd = app (bad app)
  where app : Bad → ⊥
        app (bad f) = f (bad f)

-- _≤_: Less-or-equal predicate

data  _≤_ : ℕ → ℕ → Set where
  z≤n : {n : ℕ} → zero  ≤ n                       
  s≤s : {m : ℕ} → {n : ℕ} →   m ≤ n  →  suc m ≤ suc n

infix 4 _≤_

-- Some proofs ...

4≰3 : 4 ≤ 3 → ⊥
4≰3 (s≤s (s≤s (s≤s ()))) = absurd
```

This error occurs when the compiler manages the defintion `4≰3 (s≤s (s≤s (s≤s ()))) = absurd`.

```shell
An internal error has occurred. Please report this as a bug.
Location of the error: src/full/Agda/TypeChecking/Serialise/Instances/Internal.hs:147
```

**Comment by uncle-betty** [comment]
A clause that pattern-matches with the absurd pattern `()`, doesn't have a right-hand side. And, indeed, removing the right-hand side (`= absurd`) from the `4≰3` clause makes the issue go away.

However, Agda normally doesn't produce an obscure internal error in this case. Instead, it produces a helpful warning that says that you cannot have a right-hand side. So, something is wrong.

The internal error is triggered, because instead of displaying the warning and aborting, Agda tries to serialize the warning into the `.agdai` file. The `iWarnings` of the `Interface` to be serialized looks as follows. The troublemaker is the `MetaS` in the `conPType`, which causes the internal error.

```
iWarnings = [
  TCWarning {
    tcWarningFileLine = AgdaSourceErrorLocation { ... },
    tcWarningRange = Range ( ... ) (fromList [ ... ]),
    tcWarning = AbsurdPatternRequiresNoRHS [
      Arg {
        argInfo = ArgInfo { ... },
        unArg = Named {
          nameOf = Nothing,
          namedThing = ConP (ConHead { ... }) (ConPatternInfo {
            conPInfo = PatternInfo { ...  },
            conPRecord = False,
            conPFallThrough = False,
            conPType = Just (Arg {
              argInfo = ArgInfo { ... },
              unArg = El {
                _getSort = MetaS (MetaId 19) [],
                unEl = Def (QName { ... Foo._≤_ ... }) [
                  Apply ( ... 4 ... ),
                  Apply ( ... 3 ... )
                ]
              }
            }), conPLazy = False
          }) [ ... ]
        }
      }
    ],
    tcWarningPrintedWarning = /home/tlo/repro/Foo.agda:33,28-34
      The right-hand side must be omitted if there is an absurd pattern,
      () or {}, in the left-hand side.
      when checking that the clause 4≰3 (s≤s (s≤s (s≤s ()))) = absurd has
      type 4 ≤ 3 → ⊥,
    tcWarningCached = True
  }
]
```

So, I guess the question is: Why doesn't Agda print the warning and fail, and instead silently tries to serialize the warning?

Oh, and this is still an issue with current master.

**Comment by gallais** [comment]
> So, I guess the question is: Why doesn't Agda print the warning and fail, and instead silently tries to serialize the warning?

Because this is a benign warning that can be easily recovered from by dropping the needless RHS.
But clearly we are trying to serialise a bit too much when saving this warning.

**Comment by uncle-betty** [comment]
Oooh, yes, right. That's why it's called a *warning*, I guess.  :) (As opposed to an error.)

FWIW, a stripped-down MWE:

```agda
module Foo where

open import Data.Nat using (ℕ ; _≤_ ; s≤s)

foo : 2 ≤ 1 → ℕ 
foo (s≤s ()) = 123
```

Note that one needs at least one `s≤s`. The following doesn't have the issue.

```agda
module Foo where

open import Data.Nat using (ℕ ; _≤_ ; s≤s)

foo : 1 ≤ 0 → ℕ 
foo () = 123
```

**Comment by jespercockx** [comment]
Since metavariables are not serialized, there is normally a pass that turns all open metavariables into postulates before serialization (see https://github.com/agda/agda/blob/38615d640f63522ceb560f5f4d2696e53d06b007/src/full/Agda/TypeChecking/MetaVars.hs#L1478-L1527). Maybe the problem here is that one metavariable is missed by this process.

**Comment by uncle-betty** [comment]
Thanks for the pointer towards meta instantiation! It seems to have done the trick. I think that I was able to fix the issue. See #4903.

I basically followed the lead of `checkFunDefS`, which uses `instantiateFull` to get rid of metas in clauses. I figured that the same method would work for the argument information in the warning, as clauses contain that same argument information.

I don't yet fully grasp meta instantiation, though. In particular, I don't yet understand how what I did relates to the piece of code pointed out above, which creates postulates for open metas. Can I read about this somewhere?

Oh, and I'm realizing that I forgot to add a test case on #4903. I'll do that now.
